### ISSUE TITLE
Evaluate Shopify remote-ui library for rendering application content using multi-threaded webworkers.

### ISSUE THREAD

**Issue by mawburn** [foundryvtt/foundryvtt#8296]
### Performance

So, we use and created this library where I work: https://github.com/Shopify/remote-ui

What it essentially does is allows custom extensions/plugins to be multithreaded by spinning each one over into a web worker allowing the browser to put the plugin in it's own dedicated instance, thereby its own thread, since JS is single threaded within the browser. 

I know some other large companies also use this outside of the one I work for, but it's incredibly useful and robust for our large use case. I haven't worked on it directly, but I have worked with it a lot and it's pretty great honestly. However the biggest downfall of using it, is debugging can be quite complex _SOMETIMES_, but usually only if you run into a bug that is directly related to the remote-ui itself. Other than that, plugin developers won't know or care about the web worker itself. 

There might be other libraries that do this, but in theory this should increase the performance on lower end machines significantly. 

If you were to adopt it, I would suggest maybe using it on some internal built in systems that work similarly to plugins at first until you're comfortable with it. I could probably lean to offer some help if you needed it, but can't take **any kind** of compensation for it. 

You can ping me on Discord @ hypnoCode#4242 if you need clarification on this or have more questions.

**Comment by aaclayton** [comment]
Hey @mawburn thanks for putting this library on my radar. We have been starting to invest in web workers some with Foundry VTT version 10, and we intend to lean into it more in the future. It would be good to evaluate this library and decide whether its worth relying on it as opposed to implementing our own bespoke approach.

**Comment by mawburn** [comment]
Yeah, that's basically what they did. It started out as one of my coworker's side project and then blew up because it was incredibly useful. I would advise to not build your own, because it's deceptively complicated and extremely hard to debug the core library since you run into weird race conditions that don't normally happen in JS because it's not multi-threaded. Like [this when it was comparing internal vs external IDs](https://user-images.githubusercontent.com/2829924/137385827-718ef9f4-1632-4e04-92fd-1d9b5d7081ba.png) when the external IDs were unmounted but the internal got out of sync. They fixed this already.

Here's a more in-depth look at it:
https://shopify.engineering/remote-rendering-ui-extensibility

I'm pretty sure the main author did a talk on it at a conference when he initially released it. If I can find it I'll replace this line with it. 

[Here is a video on it](https://www.youtube.com/watch?v=I4ftfY7x3EQ), but it's not the one I'm thinking of but it's good.

**Comment by mawburn** [comment]
I just learned of a better documented and a library that has more traction: 
https://github.com/BuilderIO/partytown

https://partytown.builder.io/

**Comment by aaclayton** [comment]
I think at this point it's unlikely we proceed here as this doesn't really fit in to our development roadmap.
